eos-boot-upgrade

```diff
@@ -0,0 +1,267 @@
+    return disk
+
+def get_esp(disk):
+    '''Get the ESP partition data'''
+    esp_partition = disk.get_partition_by_guid('C12A7328-F81F-11D2-BA4B-00A0C93EC93B')
+    if not esp_partition:
+        raise BootUpgradeException('[!] ESP partition not found (or wrong GUID)')
+    if not esp_partition.device.endswith('1'):
+        raise BootUpgradeException('[!] ESP is not the first partition')
+    print('[*] ESP partition: ' + esp_partition.device)
+
+    return esp_partition
+
+def get_dir_size(dir_path):
+    '''Get the directory size'''
+    total_size = 0
+    for (dirpath, dirnames, filenames) in os.walk(dir_path):
+        for f in filenames:
+            fp = os.path.join(dirpath, f)
+            total_size += (os.path.getsize(fp) if os.path.isfile(fp) else 0)
+
+    return total_size
+
+def get_partition_free_space(dir_path):
+    '''Calculate the free space available in the partition'''
+    (_, _, free) = shutil.disk_usage(dir_path);
+    return free
+
+def check_esp_space(esp_mount, ostree_boot_efi_path):
+    '''Check that we have enough space in the ESP to copy the new data'''
+    ostree_boot_efi_size = get_dir_size(ostree_boot_efi_path);
+    esp_free_space = get_partition_free_space(esp_mount);
+
+    return esp_free_space > ostree_boot_efi_size
+
+def _dir_cmp(dcmp):
+    '''Helper for directories comparison'''
+    if dcmp.diff_files or dcmp.right_only:
+        return True
+    for sub_dcmp in dcmp.subdirs.values():
+        if _dir_cmp(sub_dcmp):
+            return True
+    return False
+
+def check_for_upgrade(esp_mount, ostree_boot_efi_path):
+    '''Recursively compare the files to upgrade'''
+    dcmp = filecmp.dircmp(esp_mount, ostree_boot_efi_path)
+    return _dir_cmp(dcmp)
+
+def do_uefi_upgrade(esp_mount, ostree_boot_efi_path):
+    '''Actually start the UEFI upgrade copying the files'''
+    tmp_dest_directory = os.path.join(esp_mount, 'new')
+
+    try:
+        distutils.dir_util.copy_tree(ostree_boot_efi_path, tmp_dest_directory)
+        upgrade_list = distutils.dir_util.copy_tree(ostree_boot_efi_path, esp_mount)
+        os.sync()
+    finally:
+        distutils.dir_util.remove_tree(tmp_dest_directory)
+
+    print('[*] Updated:')
+    print(*upgrade_list, sep='\n')
+
+def start_uefi_upgrade():
+    '''Start the UEFI files upgrade'''
+    esp_mount = '/boot/efi'
+    ostree_boot_efi_path = os.path.join(cmdline.options['ostree'], 'boot/efi')
+
+    if not os.path.isdir(esp_mount):
+        raise BootUpgradeException('[!] ESP mountpoint', esp_mount, 'does not exist')
+
+    try:
+        disk = get_disk();
+        esp_partition = get_esp(disk);
+
+        with mount(esp_partition.device, esp_mount) as esp_mount:
+            if not check_for_upgrade(esp_mount, ostree_boot_efi_path):
+                print('[*] No UEFI files to upgrade')
+                return
+
+            if not check_esp_space(esp_mount, ostree_boot_efi_path):
+                raise BootUpgradeException('[!] Not enough space in the ESP')
```

Comment:
Might be better to just try the upgrade:

* Copy all the files over; if there's not enough space, this will fail
* Now rename them into place; this should need no additional space, I guess?
* On success or error, clean up the temporary directory